spark_log_parser/parsing_models/application_model_v2.py

```diff
@@ -1,131 +1,505 @@
+        bucket = path[0]
+        key = ("/".join(path[1:])).lstrip("/") + ".json"
 
-        val1.validate()
+        if compress is False:
+            s3.put_object(Bucket=bucket, Body=json.dumps(saveDat).encode("utf-8"), Key=key)
+        else:
+            dat = gzip.compress(json.dumps(saveDat).encode("utf-8"))
+            s3.put_object(Bucket=bucket, Body=dat, Key=key + ".gz")
 
-        val2 = EventDataValidation(app=appobj, debug=debug)
-        val2.validate()
+        logging.info("Saved object to cloud: %s" % (key))
 
-    def getRecentEvents(self):
 
-        tcomp = np.concatenate(([0.0], self.taskData["end_time"].values))
+class AbstractLocalFileDataLoader(abc.ABC, DataLoader):
+    @abc.abstractmethod
+    def read_uncompressed_file(self, file):
+        return file.read()
 
-        if self.existsSQL:
-            tcomp = np.concatenate(
-                (tcomp, self.sqlData["start_time"].values, self.sqlData["end_time"].values)
+    @abc.abstractmethod
+    def read_compressed_file(self, file):
+        return file.read().decode("ascii")
+
+    def load_from_local(self, filepath):
+        if ".gz" in filepath:
+            with gzip.open(filepath, "r") as fin:
+                save_data = self.read_compressed_file(fin)
+        else:
+            with open(filepath, "r") as fin:
+                save_data = self.read_uncompressed_file(fin)
+
+        return save_data
+
+    async def batch_load_fn(self, keys):
+        return [self.load_from_local(filepath) for filepath in keys]
+
+
+class LocalFileBlobDataLoader(AbstractLocalFileDataLoader):
+    def read_uncompressed_file(self, file):
+        return super().read_uncompressed_file(file)
+
+    def read_compressed_file(self, file):
+        return super().read_compressed_file(file)
+
+
+class LocalFileLinesDataLoader(AbstractLocalFileDataLoader):
+    def read_uncompressed_file(self, file):
+        lines = []
+        for line in file:
+            lines.append(line)
+
+        return lines
+
+    def read_compressed_file(self, file):
+        return self.read_uncompressed_file(file)
+
+
+class AbstractS3FileDataLoader(abc.ABC, DataLoader):
+    _s3 = None
+
+    @property
+    def s3(self):
+        if self._s3 is None:
+            self._s3 = boto3.resource("s3")
+
+        return self._s3
+
+    @abc.abstractmethod
+    def read_uncompressed_data(self, data):
+        return data.decode("ascii")
+
+    @abc.abstractmethod
+    def read_compressed_data(self, data):
+        uncompressed = gzip.decompress(data)
+        return uncompressed.decode("utf-8")
+
+    def load_from_s3(self, filepath):
+        # TODO - there must be some better way of parsing this?
+        path = filepath.replace("s3://", "").split("/")
+        bucket = path[0]
+        key = ("/".join(path[1:])).lstrip("/")
+
+        data = self.s3.Object(bucket, key).get()["Body"].read()
+
+        # TODO - there must be a better way of testing if a filepath is for a compressed/gzipped file...
```

Comment:
You can check the first two bytes, which is probably a better solution. According to this thread it's not 100% reliable, but still probably better than looking for .gz. It's surprising the gzip package doesn't have a detection method like the zip package does.
https://stackoverflow.com/questions/3703276/how-to-tell-if-a-file-is-gzip-compressed

Reply:
> It's surprising the gzip package doesn't have a detection method like the zip package does

Yeah, I wanted there to be something like this! The suggestion to test the first two bytes is an interesting one - I wonder if it would be better to require the proper file extension, though, from an API contract standpoint 🤔

Reply:
I guess what I mean is - from a robustness standpoint, yes, testing the bytes would be great. However, since users are uploading these files directly to us, we may be able to be a bit more strict with the format we expect file paths/names to be in, and therefore be able to rely on those validations later on in our code. Should we really be expected to "detect" the format of a file with no extension provided? Certainly the answer would be "yes" if this was some e.g. Operating System utility, but maybe we don't need to burden ourselves with that?